File: swift/Sources/WaxAndWaneCore/Policy.swift
import Foundation

func clampUnit(_ value: Float) -> Float { min(max(value, 0.0), 1.0) }

func normalizeAmbient(_ ambient: Float, dark: Float, bright: Float, gamma: Float) -> Float {
    let linear = clampUnit((ambient - dark) / (bright - dark))
    return pow(linear, gamma)
}

func mapAmbient(_ ambient: Float, minValue: Float, maxValue: Float, invert: Bool) -> Float {
    invert ? maxValue - ambient * (maxValue - minValue)
           : minValue + ambient * (maxValue - minValue)
}

struct RingBuffer {
    private var buf: [Float]
    private var index = 0
    private var count = 0
    private var total: Float = 0
    let capacity: Int

    init(capacity: Int) {
        precondition(capacity > 0, "RingBuffer capacity must be > 0")
        self.capacity = capacity
        self.buf = [Float](repeating: 0, count: capacity)
    }

    mutating func append(_ value: Float) {
        if count == capacity { total -= buf[index] } else { count += 1 }
        buf[index] = value
        total += value
        index = (index + 1) % capacity
    }

    var mean: Float { count > 0 ? total / Float(count) : 0 }
    var isEmpty: Bool { count == 0 }
}

func thresholdForDelta(_ delta: Float, changeThreshold: Float, riseThreshold: Float?, fallThreshold: Float?) -> Float {
    if delta > 0 { return riseThreshold ?? changeThreshold }
    if delta < 0 { return fallThreshold ?? changeThreshold }
    return changeThreshold
}

struct OutputChannelSettings {
    let control: BrightnessControl
    let minValue: Float
    let maxValue: Float
    let invert: Bool
    let manualValue: Float
}

func targetForControl(
    control: BrightnessControl,
    smoothedAmbient: Float,
    lastValue: Float,
    minValue: Float,
    maxValue: Float,
    invert: Bool,
    manualValue: Float,
    changeThreshold: Float,
    riseThreshold: Float? = nil,
    fallThreshold: Float? = nil
) -> Float? {
    targetForChannel(
        OutputChannelSettings(
            control: control,
            minValue: minValue,
            maxValue: maxValue,
            invert: invert,
            manualValue: manualValue
        ),
        smoothedAmbient: smoothedAmbient,
        lastValue: lastValue,
        changeThreshold: changeThreshold,
        riseThreshold: riseThreshold,
        fallThreshold: fallThreshold
    )
}

func targetForChannel(
    _ channel: OutputChannelSettings,
    smoothedAmbient: Float,
    lastValue: Float,
    changeThreshold: Float,
    riseThreshold: Float? = nil,
    fallThreshold: Float? = nil
) -> Float? {
    let target: Float
    switch channel.control {
    case .system:
        return nil
    case .manual:
        target = channel.manualValue
    case .auto:
        target = mapAmbient(smoothedAmbient, minValue: channel.minValue, maxValue: channel.maxValue, invert: channel.invert)
    }
    let delta = target - lastValue
    return abs(delta) > thresholdForDelta(delta, changeThreshold: changeThreshold, riseThreshold: riseThreshold, fallThreshold: fallThreshold) ? target : nil
}

func computeTargets(
    history: inout RingBuffer,
    ambientNow: Float,
    lastKeyboard: Float,
    lastScreen: Float,
    s: Settings
) -> (keyboard: Float?, screen: Float?) {
    history.append(ambientNow)
    let smoothedRaw = history.mean
    let calibrated = normalizeAmbient(smoothedRaw, dark: s.ambientDark, bright: s.ambientBright, gamma: s.outputGamma)

    let keyboard = OutputChannelSettings(
        control: s.keyboardControl,
        minValue: s.keyboardMin,
        maxValue: s.keyboardMax,
        invert: s.invertKeyboard,
        manualValue: s.manualKeyboardBrightness
    )
    let screen = OutputChannelSettings(
        control: s.screenControl,
        minValue: s.screenMin,
        maxValue: s.screenMax,
        invert: s.invertScreen,
        manualValue: s.manualScreenBrightness
    )

    return (
        targetForChannel(
            keyboard,
            smoothedAmbient: calibrated,
            lastValue: lastKeyboard,
            changeThreshold: s.changeThreshold,
            riseThreshold: s.riseThreshold,
            fallThreshold: s.fallThreshold
        ),
        targetForChannel(
            screen,
            smoothedAmbient: calibrated,
            lastValue: lastScreen,
            changeThreshold: s.changeThreshold,
            riseThreshold: s.riseThreshold,
            fallThreshold: s.fallThreshold
        )
    )
}

File: python/Sources/policy.py
#!/usr/bin/env python3
"""Policy module for Wax and Wane.

Contains pure functions for brightness calculation and mapping logic.
No I/O operations - suitable for unit testing.
"""

from typing import Optional, Tuple
from collections import deque
from dataclasses import dataclass


@dataclass(frozen=True)
class OutputChannelSettings:
    """Settings needed to compute one brightness channel target."""
    control: str
    minimum: float
    maximum: float
    invert: bool
    manual_value: float


def normalize_ambient(ambient: float, dark: float, bright: float, gamma: float) -> float:
    """Normalize ambient light value to [0, 1] range with gamma correction.
    
    Args:
        ambient: Raw ambient light measurement.
        dark: Calibration point for minimum ambient light.
        bright: Calibration point for maximum ambient light.
        gamma: Gamma correction factor (must be > 0 and <= 10).
        
    Returns:
        Normalized value in [0, 1] range with gamma applied.
    """
    if bright == dark:
        return 0.5 ** gamma
    linear = min(max((ambient - dark) / (bright - dark), 0.0), 1.0)
    return linear ** gamma


def map_ambient(ambient: float, out_min: float, out_max: float, invert: bool) -> float:
    """Map normalized ambient value to output range.
    
    Args:
        ambient: Normalized ambient light value [0, 1].
        out_min: Minimum output value.
        out_max: Maximum output value.
        invert: If True, invert the mapping (bright→dark).
        
    Returns:
        Mapped output value.
    """
    if invert:
        return out_max - ambient * (out_max - out_min)
    return out_min + ambient * (out_max - out_min)


def _threshold_for_delta(delta: float, change_threshold: float, 
                         rise_threshold: Optional[float], 
                         fall_threshold: Optional[float]) -> float:
    """Determine appropriate threshold based on delta direction.
    
    Args:
        delta: Change in brightness value.
        change_threshold: Default threshold for both directions.
        rise_threshold: Optional separate threshold for increases.
        fall_threshold: Optional separate threshold for decreases.
        
    Returns:
        Appropriate threshold value.
    """
    if delta > 0:
        return change_threshold if rise_threshold is None else rise_threshold
    if delta < 0:
        return change_threshold if fall_threshold is None else fall_threshold
    return change_threshold


def target_for_channel(
    channel: OutputChannelSettings,
    smoothed_ambient: float,
    last_value: float,
    change_threshold: float,
    rise_threshold: Optional[float] = None,
    fall_threshold: Optional[float] = None,
) -> Optional[float]:
    """Calculate target brightness for one output channel.
    
    Args:
        channel: Channel configuration settings.
        smoothed_ambient: Smoothed and calibrated ambient light value.
        last_value: Previous brightness setting for this channel.
        change_threshold: Minimum change to trigger update.
        rise_threshold: Optional separate threshold for brightness increases.
        fall_threshold: Optional separate threshold for brightness decreases.
        
    Returns:
        Target brightness value, or None if no change needed or channel is system-controlled.
    """
    if channel.control == "system":
        return None
    if channel.control == "manual":
        target = channel.manual_value
    elif channel.control == "auto":
        target = map_ambient(smoothed_ambient, channel.minimum, channel.maximum, channel.invert)
    else:
        raise ValueError(f"Unsupported brightness control mode: {channel.control}")

    delta = target - last_value
    threshold = _threshold_for_delta(delta, change_threshold, rise_threshold, fall_threshold)
    return target if abs(delta) > threshold else None


def target_for_control(
    control: str,
    smoothed_ambient: float,
    last_value: float,
    minimum: float,
    maximum: float,
    invert: bool,
    manual_value: float,
    change_threshold: float,
    rise_threshold: Optional[float] = None,
    fall_threshold: Optional[float] = None,
) -> Optional[float]:
    """Legacy wrapper for target_for_channel.
    
    Args:
        control: Control mode ("auto", "manual", "system").
        smoothed_ambient: Smoothed ambient light value.
        last_value: Previous brightness value.
        minimum: Minimum output brightness.
        maximum: Maximum output brightness.
        invert: Whether to invert the mapping.
        manual_value: Fixed brightness for manual mode.
        change_threshold: Minimum change threshold.
        rise_threshold: Optional rise-specific threshold.
        fall_threshold: Optional fall-specific threshold.
        
    Returns:
        Target brightness or None.
    """
    return target_for_channel(
        OutputChannelSettings(control, minimum, maximum, invert, manual_value),
        smoothed_ambient,
        last_value,
        change_threshold,
        rise_threshold,
        fall_threshold,
    )


def compute_targets(
    history: deque,
    ambient_now: float,
    last_keyboard: float,
    last_screen: float,
    settings,
) -> Tuple[Optional[float], Optional[float]]:
    """Compute target brightness values for keyboard and screen.
    
    This is a pure function - it performs no I/O and does not mutate history.
    The caller must append ambient_now to history before calling this function.
    
    Args:
        history: Deque of recent ambient light measurements.
        ambient_now: Current ambient light measurement.
        last_keyboard: Previous keyboard brightness value.
        last_screen: Previous screen brightness value.
        settings: Settings object with configuration.
        
    Returns:
        Tuple of (keyboard_target, screen_target), each may be None.
    """
    smoothed = sum(history) / len(history) if history else ambient_now
    calibrated = normalize_ambient(
        smoothed, 
        settings.ambient_dark, 
        settings.ambient_bright, 
        settings.output_gamma
    )

    channels = (
        OutputChannelSettings(
            settings.keyboard_control, 
            settings.keyboard_min, 
            settings.keyboard_max,
            settings.invert_keyboard, 
            settings.manual_keyboard_brightness,
        ),
        OutputChannelSettings(
            settings.screen_control, 
            settings.screen_min, 
            settings.screen_max,
            settings.invert_screen, 
            settings.manual_screen_brightness,
        ),
    )
    return tuple(
        target_for_channel(
            channel, 
            calibrated, 
            last_value, 
            settings.change_threshold,
            settings.rise_threshold, 
            settings.fall_threshold,
        )
        for channel, last_value in zip(channels, (last_keyboard, last_screen))
    )
